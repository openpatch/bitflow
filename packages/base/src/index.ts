export * from "./findLast";
export * from "./groupBy";
export * from "./lerpColor";
export * from "./levenshtein";
export * from "./schemas";
export * from "./uuid";

import { Input, Privacy, Task, TaskAnswer, TaskResult, Title } from "./schemas";

export interface Action {
  type: string;
  payload?: any;
}

export interface TitleProps<T extends Title> {
  title: Pick<T, "view" | "subtype">;
}

export interface TitleViewFormProps {
  name: string;
}

export interface PrivacyProps<P extends Privacy> {
  privacy: Pick<P, "view" | "subtype">;
}
export interface PrivacyViewFormProps {
  name: string;
}

export interface InputProps<I extends Input> {
  input: Pick<I, "view" | "subtype">;
}

export interface InputViewFormProps {
  name: string;
}

export interface TaskProps<
  T extends Task,
  R extends TaskResult,
  A extends TaskAnswer,
  Act extends Action
> {
  mode: "default" | "recording" | "result";
  task: Pick<T, "view" | "subtype">;
  result?: R;
  answer?: A;
  onChange?: (value: A) => void;
  onAction?: (action: Act) => void;
}
export type TaskRef<Act> = {
  dispatch: (action: Act) => void;
};

export interface TaskEvaluationFormProps {
  name: string;
}

export interface TaskViewFormProps {
  name: string;
}

export interface TaskFeedbackFormProps {
  name: string;
}

export interface TaskStatistic {
  count: number;
}

export interface TaskStatisticProps<S extends TaskStatistic, T extends Task> {
  statistic: S;
  task: T;
}

export interface TaskEvaluateParams<A extends TaskAnswer, T extends Task> {
  answer?: A;
  task: T;
}
export type Evaluate<
  A extends TaskAnswer,
  T extends Task,
  R extends TaskResult
> = ({ answer, task }: TaskEvaluateParams<A, T>) => Promise<R>;

export interface UpdateTaskStatisticParams<
  S extends TaskStatistic,
  A extends TaskAnswer,
  T extends Task,
  R extends TaskResult
> {
  statistic?: S;
  answer: A;
  task: T;
  result?: R;
}
export type UpdateTaskStatistic<
  S extends TaskStatistic,
  A extends TaskAnswer,
  T extends Task,
  R extends TaskResult
> = ({
  statistic,
  answer,
  task,
  result,
}: UpdateTaskStatisticParams<S, A, T, R>) => Promise<S>;
