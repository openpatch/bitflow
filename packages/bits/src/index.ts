export * from "@openpatch/bits-base";
export * from "@openpatch/bits-input-markdown";
export * from "@openpatch/bits-privacy-markdown";
export * from "@openpatch/bits-shell";
export * from "@openpatch/bits-task-choice";
