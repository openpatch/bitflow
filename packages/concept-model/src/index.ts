export * from "./ConceptModel";
export * from "./LatentVariableNode";
export * from "./ManifestTaskNode";
