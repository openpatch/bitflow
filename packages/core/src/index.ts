export * from "./bits";
export * from "./bitsSchema";
export * from "./do";
export * from "./doSchema";
export * from "./findLast";
export * from "./flow";
export * from "./flowSchema";
export * from "./groupBy";
export * from "./lerpColor";
export * from "./levenshtein";
export * from "./uuid";
