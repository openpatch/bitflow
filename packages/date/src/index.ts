export * from "./date";
export * from "./useDate";
