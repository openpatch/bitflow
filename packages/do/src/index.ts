export * from "./Do";
export * from "./DoCheckpoint";
export * from "./DoEnd";
export * from "./DoInput";
export * from "./DoStart";
export * from "./DoSynchronize";
export * from "./DoTask";
export * from "./DoTitle";
export * from "./types";
