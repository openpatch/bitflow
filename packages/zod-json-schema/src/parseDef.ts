import { ZodFirstPartyTypeKind, ZodLazy, ZodSchema } from "zod";
import { JsonSchema7ArrayType, parseArrayDef } from "./parsers/array";
import { JsonSchema7BigintType, parseBigintDef } from "./parsers/bigint";
import { JsonSchema7BooleanType, parseBooleanDef } from "./parsers/boolean";
import { JsonSchema7DateType, parseDateDef } from "./parsers/date";
import { JsonSchema7EnumType, parseEnumDef } from "./parsers/enum";
import { parseIntersectionDef } from "./parsers/intersection";
import { JsonSchema7LiteralType, parseLiteralDef } from "./parsers/literal";
import { JsonSchema7MapType, parseMapDef } from "./parsers/map";
import {
  JsonSchema7NativeEnumType,
  parseNativeEnumDef,
} from "./parsers/nativeEnum";
import { JsonSchema7NullType, parseNullDef } from "./parsers/null";
import { JsonSchema7NullableType, parseNullableDef } from "./parsers/nullable";
import { JsonSchema7NumberType, parseNumberDef } from "./parsers/number";
import { JsonSchema7ObjectType, parseObjectDef } from "./parsers/object";
import { JsonSchema7RecordType, parseRecordDef } from "./parsers/record";
import { parseSetDef } from "./parsers/set";
import { JsonSchema7StringType, parseStringDef } from "./parsers/string";
import { JsonSchema7TupleType, parseTupleDef } from "./parsers/tuple";
import {
  JsonSchema7UndefinedType,
  parseUndefinedDef,
} from "./parsers/undefined";
import {
  JsonSchema7AnyOfType,
  JsonSchema7PrimitiveUnionType,
  parseUnionDef,
} from "./parsers/union";

type JsonSchema7AnyType = {};
type JsonSchema7RefType = { $ref: string };

export type JsonSchema7Type =
  | JsonSchema7StringType
  | JsonSchema7ArrayType
  | JsonSchema7NumberType
  | JsonSchema7BigintType
  | JsonSchema7BooleanType
  | JsonSchema7DateType
  | JsonSchema7EnumType
  | JsonSchema7LiteralType
  | JsonSchema7NativeEnumType
  | JsonSchema7NullType
  | JsonSchema7NumberType
  | JsonSchema7ObjectType
  | JsonSchema7RecordType
  | JsonSchema7TupleType
  | JsonSchema7PrimitiveUnionType
  | JsonSchema7UndefinedType
  | JsonSchema7AnyOfType
  | JsonSchema7RefType
  | JsonSchema7AnyType
  | JsonSchema7MapType
  | JsonSchema7NullableType;

export type Visited = { schema: ZodSchema<any>; path: string[] }[];

export function parseDef<T>(
  schema: ZodSchema<any>,
  path: string[],
  visited: Visited
): JsonSchema7Type | undefined {
  const wasVisited = visited.find((x) => Object.is(x.schema, schema));
  if (wasVisited) {
    return { $ref: `#/${wasVisited.path.join("/")}` };
  } else {
    visited.push({ schema, path });
  }

  const def = schema._def as any;
  const typeName: ZodFirstPartyTypeKind = def.typeName as ZodFirstPartyTypeKind;

  switch (typeName) {
    case ZodFirstPartyTypeKind.ZodString:
      return parseStringDef(def);
    case ZodFirstPartyTypeKind.ZodNumber:
      return parseNumberDef(def);
    case ZodFirstPartyTypeKind.ZodObject:
      return parseObjectDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodBigInt:
      return parseBigintDef(def);
    case ZodFirstPartyTypeKind.ZodBoolean:
      return parseBooleanDef();
    case ZodFirstPartyTypeKind.ZodDate:
      return parseDateDef();
    case ZodFirstPartyTypeKind.ZodUndefined:
      return parseUndefinedDef();
    case ZodFirstPartyTypeKind.ZodNull:
      return parseNullDef();
    case ZodFirstPartyTypeKind.ZodArray:
      return parseArrayDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodUnion:
      return parseUnionDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodIntersection:
      return parseIntersectionDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodTuple:
      return parseTupleDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodRecord:
      return parseRecordDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodLiteral:
      return parseLiteralDef(def);
    case ZodFirstPartyTypeKind.ZodEnum:
      return parseEnumDef(def);
    case ZodFirstPartyTypeKind.ZodNativeEnum:
      return parseNativeEnumDef(def);
    case ZodFirstPartyTypeKind.ZodNullable:
      return parseNullableDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodOptional:
      return parseDef(def.innerType, path, visited);
    case ZodFirstPartyTypeKind.ZodMap:
      return parseMapDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodSet:
      return parseSetDef(def, path, visited);
    case ZodFirstPartyTypeKind.ZodEffects:
    case ZodFirstPartyTypeKind.ZodAny:
    case ZodFirstPartyTypeKind.ZodUnknown:
    case ZodFirstPartyTypeKind.ZodDefault:
      return {};
    case ZodFirstPartyTypeKind.ZodLazy:
      return parseDef((schema as ZodLazy<any>)._def.getter(), path, visited);
    case ZodFirstPartyTypeKind.ZodNever:
    case ZodFirstPartyTypeKind.ZodFunction:
    case ZodFirstPartyTypeKind.ZodPromise:
    case ZodFirstPartyTypeKind.ZodVoid:
      return undefined;
    default:
      return ((_: never) => undefined)(typeName);
  }
}
