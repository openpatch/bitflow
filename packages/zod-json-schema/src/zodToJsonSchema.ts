import { ZodSchema } from "zod";
import { JsonSchema7Type, parseDef } from "./parseDef";

/**
 * @param schema The Zod schema to be converted
 */
function zodToJsonSchema(schema: ZodSchema<any>):
  | ({
      $schema: "http://json-schema.org/draft-07/schema#";
    } & JsonSchema7Type)
  | unknown;
/**
 * @param schema The Zod schema to be converted
 * @param name The (optional) name of the schema. If a name is passed, the schema will be put in 'definitions' and referenced from the root.
 */
function zodToJsonSchema<T extends string>(
  schema: ZodSchema<any>,
  name: T
): {
  $schema: "http://json-schema.org/draft-07/schema#";
  $ref: string;
  definitions: Record<T, JsonSchema7Type>;
};
function zodToJsonSchema(schema: ZodSchema<any>, name?: string) {
  // console.log(schema)

  return name === undefined
    ? {
        $schema: "http://json-schema.org/draft-07/schema#",
        ...parseDef(schema, [], []),
      }
    : {
        $schema: "http://json-schema.org/draft-07/schema#",
        $ref: `#/$defs/${name}`,
        $defs: {
          [name]: parseDef(schema, ["$defs", name], []) || {},
        },
      };
}

export { zodToJsonSchema };
