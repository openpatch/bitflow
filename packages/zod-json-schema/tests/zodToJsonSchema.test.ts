import { z } from "zod";
import { zodToJsonSchema } from "../src/zodToJsonSchema";

describe("Root schema result after parsing", () => {
  it("should return the schema directly in the root if no name is passed", () => {
    expect(zodToJsonSchema(z.any())).toStrictEqual({
      $schema: "http://json-schema.org/draft-07/schema#",
    });
  });
  it('should return the schema inside a named property in "definitions" if a name is passed', () => {
    expect(zodToJsonSchema(z.any(), "MySchema")).toStrictEqual({
      $schema: "http://json-schema.org/draft-07/schema#",
      $ref: `#/$defs/MySchema`,
      $defs: {
        MySchema: {},
      },
    });
  });
});
