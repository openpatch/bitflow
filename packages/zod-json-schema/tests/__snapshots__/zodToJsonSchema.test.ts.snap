// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitflow schema shoudl return the schema 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": Object {
    "description": Object {
      "type": "string",
    },
    "name": Object {
      "type": "string",
    },
    "subtype": Object {
      "const": "simple",
      "type": "string",
    },
    "view": Object {
      "additionalProperties": false,
      "properties": Object {
        "message": Object {
          "type": "string",
        },
        "title": Object {
          "type": "string",
        },
      },
      "required": Array [
        "title",
        "message",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "name",
    "description",
    "subtype",
    "view",
  ],
  "type": "object",
}
`;
